### PR TITLE
Add helm chart dependency tracking

### PR DESCRIPTION
**Pull request by sldblog**
## What does this pull request do?

Add helm chart dependency tracking to https://structurizr.com/share/56937/documentation/*#2

The main changes are in 2de6aa01cad5b69cd33d2e0918919d7c5368eee2

## What is the intent behind these changes?

To expose how up to date services are with the shared libraries/charts we provide

### CODE REVIEW

**Comment by sldblog** [comment]
For me, it generates this:

## Dependencies

- Latest [CircleCI orb](https://circleci.com/developer/orbs/orb/ministryofjustice/hmpps)
- Latest [gradle-spring-boot](https://plugins.gradle.org/plugin/uk.gov.justice.hmpps.gradle-spring-boot)
- Latest [helm charts](https://github.com/ministryofjustice/hmpps-helm-charts/releases)

| Software System | Application | CircleCI orb versions | gradle-spring-boot version | helm charts
| --- | --- | --- | --- | --- |
| Assess Risks And Needs| [Assessment Service](https://github.com/ministryofjustice/hmpps-assessments-api)| hmpps@3.11| 3.3.8| no dependencies|
| Assess Risks And Needs| [Risk Assessment UI](https://github.com/ministryofjustice/hmpps-risk-assessment-ui)| hmpps@3.2| | generic-prometheus-alerts@0.1.3|
| Assess Risks And Needs| [Risks and Needs Service](https://github.com/ministryofjustice/hmpps-assess-risks-and-needs)| hmpps@3.10| 3.3.8| generic-service@1.0.10<br>generic-prometheus-alerts@0.1.4|
| Book Video Link| [Book video link service](https://github.com/ministryofjustice/hmpps-book-video-link)| hmpps@3.11| | generic-prometheus-alerts@0.1.3|
| Complexity of Need Microservice| [Complexity of Need Microservice](https://github.com/ministryofjustice/hmpps-complexity-of-need)| | | no helm chart|
| Court register| [API](https://github.com/ministryofjustice/court-register)| hmpps@3.11| 3.3.13-beta-2| generic-service@1.0.16<br>generic-prometheus-alerts@0.1.5|
| Curious| [Curious API](https://github.com/ministryofjustice/curious-API)| | | no helm chart|
| Digital Prison Services| [Digital Prison Services](https://github.com/ministryofjustice/digital-prison-services)| hmpps@3.11| | generic-prometheus-alerts@0.1.2|
| Digital Prison Services| [Manage Key Workers](https://github.com/ministryofjustice/manage-key-workers)| hmpps@3.11| | generic-service@1.0.16<br>generic-prometheus-alerts@0.1.4|
| Digital Prison Services| [Manage Restricted Patients](https://github.com/ministryofjustice/hmpps-restricted-patients)| hmpps@3.11| | generic-prometheus-alerts@0.1.2|
| HMPPS Auth| [HMPPS Auth](https://github.com/ministryofjustice/hmpps-auth)| hmpps@3.0| 3.1.6| no dependencies|
| Keyworker API| [API](https://github.com/ministryofjustice/keyworker-api)| hmpps@3.11| 3.3.12| generic-prometheus-alerts@0.1.2|
| Manage A Supervision| [Manage A Supervision UI](https://github.com/ministryofjustice/hmpps-manage-supervisions)| hmpps@3.11| | generic-service@1.0.18<br>generic-prometheus-alerts@0.2.5|
| Manage POM Cases| [Offender Management Allocation Manager](https://github.com/ministryofjustice/offender-management-allocation-manager)| | | no helm chart|
| Manage Recalls| [Manage Recalls API](https://github.com/ministryofjustice/manage-recalls-api)| hmpps@3.11| 3.3.13| no dependencies|
| Manage Recalls| [Manage Recalls Web Application](https://github.com/ministryofjustice/manage-recalls-ui)| hmpps@3.11| | generic-service@1.1.2<br>generic-prometheus-alerts@0.2.5|
| NDelius| [Community API](https://github.com/ministryofjustice/community-api)| hmpps@3.11| 3.3.12| no helm chart|
| NDelius| [Delius API](https://github.com/ministryofjustice/hmpps-delius-api)| hmpps@3.11.10| 3.3.12| no helm chart|
| NDelius| [Probation Offender Events](https://github.com/ministryofjustice/probation-offender-events)| hmpps@3.11| 3.3.12| generic-service@1.0.16<br>generic-prometheus-alerts@0.1.5|
| NDelius| [Probation Offender Search API](https://github.com/ministryofjustice/probation-offender-search)| hmpps@3.11| 3.3.12| generic-service@1.0.17<br>generic-prometheus-alerts@0.2.5|
| NDelius| [Probation Offender Search Indexer](https://github.com/ministryofjustice/probation-offender-search-indexer)| hmpps@3.11| 3.3.12| generic-service@1.0.17<br>generic-prometheus-alerts@0.2.5|
| NDelius| [Probation Offender Search Webapp / PSR (New Tech)](https://github.com/ministryofjustice/ndelius-new-tech)| hmpps@3.2| 3.3.0| generic-prometheus-alerts@0.1.2|
| NOMIS| [Custody API (Deprecated)](https://github.com/ministryofjustice/custody-api)| hmpps@1.0.0| 0.4.2| no dependencies|
| NOMIS| [Prison API](https://github.com/ministryofjustice/prison-api)| hmpps@3.11| 3.3.13| generic-service@1.0.17|
| NOMIS| [Prison Offender Events](https://github.com/ministryofjustice/prison-offender-events)| hmpps@3.11| 3.3.13| generic-service@1.0.16<br>generic-prometheus-alerts@0.2.7|
| NOMIS| [PrisonerSearch](https://github.com/ministryofjustice/prisoner-offender-search)| hmpps@3.11| 3.3.13| generic-service@1.0.16<br>generic-prometheus-alerts@0.1.5|
| OASys| [Offender Assessment Events](https://github.com/ministryofjustice/offender-assessments-events)| hmpps@3.11| 3.3.5| no dependencies|
| OASys| [Offender Assessments API](https://github.com/ministryofjustice/offender-assessments-api-kotlin)| hmpps@3.11| 3.3.8| no dependencies|
| Prepare a Case for Sentence| [Court Case Matcher](https://github.com/ministryofjustice/court-case-matcher)| hmpps@3.11.11<br>dps@1.2.1| | generic-service@1.1.2<br>generic-prometheus-alerts@0.2.7|
| Prepare a Case for Sentence| [Court Case Service](https://github.com/ministryofjustice/court-case-service)| hmpps@3.2| 3.1.1| generic-prometheus-alerts@0.1.2|
| Prepare a Case for Sentence| [Crime Portal Gateway](https://github.com/ministryofjustice/crime-portal-gateway)| hmpps@2.1.0| 3.3.12| no dependencies|
| Prepare a Case for Sentence| [Prepare a Case UI](https://github.com/ministryofjustice/prepare-a-case)| | | no dependencies|
| Prison register| [API](https://github.com/ministryofjustice/prison-register)| hmpps@3.11| 3.3.13| generic-service@1.0.16<br>generic-prometheus-alerts@0.2.5|
| Prison Visits Booking| [Backend](https://github.com/ministryofjustice/prison-visits-2)| | | no helm chart|
| Prison Visits Booking| [Frontend](https://github.com/ministryofjustice/prison-visits-public)| | | no helm chart|
| Prisoner Content Hub| [Prisoner Content Hub CMS](https://github.com/ministryofjustice/prisoner-content-hub-backend)| | | no dependencies|
| Prisoner Content Hub| [Prisoner Content Hub frontend](https://github.com/ministryofjustice/prisoner-content-hub-frontend)| hmpps@3.11| | no dependencies|
| Prisoner Money| [Prisoner Money API](https://github.com/ministryofjustice/money-to-prisoners-api)| | | no helm chart|
| Prisoner Money| [Prisoner Money Bank Admin](https://github.com/ministryofjustice/money-to-prisoners-bank-admin)| | | no helm chart|
| Prisoner Money| [Prisoner Money Digital Cashbook](https://github.com/ministryofjustice/money-to-prisoners-cashbook)| | | no helm chart|
| Prisoner Money| [Prisoner Money Intelligence Tool Service](https://github.com/ministryofjustice/money-to-prisoners-noms-ops)| | | no helm chart|
| Prisoner Money| [Prisoner Money Send Money to a Prisoner Service](https://github.com/ministryofjustice/money-to-prisoners-send-money)| | | no helm chart|
| Prisoner Money| [Prisoner Money Transaction Uploader](https://github.com/ministryofjustice/money-to-prisoners-transaction-uploader)| | | no helm chart|
| Probation team contact and reference service| [API](https://github.com/ministryofjustice/probation-teams)| hmpps@3.11| 3.3.12| generic-service@1.1.2<br>generic-prometheus-alerts@0.2.7|
| Refer and monitor an intervention| [Intervention service](https://github.com/ministryofjustice/hmpps-interventions-service)| hmpps@3.11| 3.3.12| no dependencies|
| Refer and monitor an intervention| [Intervention UI](https://github.com/ministryofjustice/hmpps-interventions-ui)| | | no helm chart|
| Restricted Patients API| [API](https://github.com/ministryofjustice/hmpps-restricted-patients-api)| hmpps@3.11| 3.3.12| generic-service@1.0.16<br>generic-prometheus-alerts@0.1.4|
| Tier Service| [Tier Service](https://github.com/ministryofjustice/hmpps-tier)| hmpps@3.11| 3.3.13-beta-2| generic-service@1.0.18<br>generic-prometheus-alerts@0.1.5|
| Tier Service| [Tier SQS Tool](https://github.com/ministryofjustice/hmpps-tier-sqs-tool)| hmpps@3.11| 3.3.12| no dependencies|
| Tier Service| [Tier to Delius Update](https://github.com/ministryofjustice/hmpps-tier-to-delius-update)| hmpps@3.11| 3.3.13-beta-2| generic-service@1.0.16<br>generic-prometheus-alerts@0.1.5|
| Token verification API| [API](https://github.com/ministryofjustice/token-verification-api)| hmpps@3.11| 3.3.13| generic-service@1.0.16<br>generic-prometheus-alerts@0.2.5|
| User Preference API| [API](https://github.com/ministryofjustice/hmpps-user-preferences)| hmpps@2.2| 3.3.13| no dependencies|
| Whereabouts API| [API](https://github.com/ministryofjustice/whereabouts-api)| hmpps@3.11| 3.3.12| generic-prometheus-alerts@0.1.2|

File: src/main/kotlin/uk/gov/justice/hmpps/architecture/documentation/versionTracker.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+private fun readHelmChartDependencies(repo: File): String {
+  val chartFile = repo.resolve("helm_deploy")
+    .walkTopDown().find { it.name == "Chart.yaml" }
+    ?: return "no helm chart"
+
+  val dependencies = ObjectMapper(YAMLFactory()).readTree(chartFile).get("dependencies")
+    ?: return "standalone chart"
+
+  return dependencies.joinToString("<br>") { "${it.get("name").textValue()}@${it.get("version").textValue()}" }
```
Comment (mxco86) [inline src/main/kotlin/uk/gov/justice/hmpps/architecture/documentation/versionTracker.kt:63]: Might be worth adding a bullet to each dependency?

**Review by mxco86 — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/hmpps/architecture/documentation/versionTracker.kt (hunk 1 above)
Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/hmpps/architecture/documentation/versionTracker.kt:63]: Will try

Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/hmpps/architecture/documentation/versionTracker.kt:63]: Well, we'll have to -- structurizr ignores the <br> tags
image

**Pull request merged**